compiler/x/amd64/objectfmt/OMRJitCodeRXObjectFormat.cpp

```diff
@@ -0,0 +1,187 @@
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "codegen/CodeGenerator.hpp"
+#include "codegen/MemoryReference.hpp"
+#include "codegen/X86Instruction.hpp"
+#include "compile/Compilation.hpp"
+#include "il/MethodSymbol.hpp"
+#include "il/SymbolReference.hpp"
+#include "infra/Assert.hpp"
+#include "objectfmt/FunctionCallData.hpp"
+#include "objectfmt/JitCodeRXObjectFormat.hpp"
+#include "runtime/Runtime.hpp"
+
+
+TR::Instruction *
+OMR::X86::AMD64::JitCodeRXObjectFormat::emitFunctionCall(TR::FunctionCallData &data)
+   {
+   TR::SymbolReference *methodSymRef;
+
+   if (data.runtimeHelperIndex > 0)
+      {
+      methodSymRef = data.cg->symRefTab()->findOrCreateRuntimeHelper(data.runtimeHelperIndex);
+      }
+   else
+      {
+      methodSymRef = data.methodSymRef;
+      }
+
+   /**
+    * If a targetAddress has been provided then use that to override whatever would be
+    * derived from the symbol reference.
+    */
+   TR_ASSERT_FATAL(!(data.runtimeHelperIndex && data.targetAddress), "a runtime helper and target address cannot both be provided");
+
+   uintptr_t targetAddress = data.targetAddress ? data.targetAddress : (uintptr_t)methodSymRef->getMethodAddress();
+
+   /**
+    * The targetAddress must be known at this point.  The only exception is a recursive
+    * call to the method being compiled, as that won't be known until binary encoding
+    * has completed.
+    */
+   TR_ASSERT_FATAL(targetAddress ||
+                   (!targetAddress && data.cg->comp()->isRecursiveMethodTarget(methodSymRef->getSymbol())), "function address is unknown");
+
+   data.cg->resetIsLeafMethod();
+
+   TR::Compilation *comp = data.cg->comp();
+
+   /**
+    * If this function won't be recompiled then allow a recursive call to
+    * be done directly with a relative displacement.
+    */
+   if (comp->isRecursiveMethodTarget(methodSymRef->getSymbol()) &&
+       !comp->couldBeRecompiled())
+      {
+      TR::X86ImmSymInstruction *callImmSymInstr;
+
+      if (data.prevInstr)
+         {
+         callImmSymInstr = generateImmSymInstruction(data.prevInstr, CALLImm4, 0, data.methodSymRef, data.regDeps, data.cg);
+         }
+      else
+         {
+         callImmSymInstr = generateImmSymInstruction(CALLImm4, data.callNode, 0, data.methodSymRef, data.regDeps, data.cg);
+         }
+
+      if (data.adjustsFramePointerBy != 0)
+         {
+         callImmSymInstr->setAdjustsFramePointerBy(data.adjustsFramePointerBy);
+         }
+
+      data.out_callInstr = callImmSymInstr;
+      return callImmSymInstr;
+      }
+
+   ccFunctionData *ccFunctionDataAddress =
+      reinterpret_cast<ccFunctionData *>(data.cg->allocateCodeMemory(sizeof(ccFunctionData), false));
```

Comment:
**Note to reviewers:**  this implementation uses `allocateCodeMemory`.   This is only a temporary solution until #5634 is merged, in which case it will switch to allocating such memory from CCData.